File: docs/code/java/src/main/java/cancellation/CancellationExample.java
package cancellation;

import dev.restate.sdk.Context;
import dev.restate.sdk.common.TerminalException;

public class CancellationExample {
  public void processOrder(Context ctx, Order order) {
    // This executes - not an await point
    ctx.set("status", "processing");

    try {
      // If cancelled before this await, run won't execute
      // If cancelled during execution, then gets thrown after the execution
      Payment payment = ctx.run(() -> processPayment(order));

      // These one-way calls execute even if cancellation happened right before
      NotificationServiceClient.fromContext(ctx)
          .send()
          .notify(order.getUserId(), "Payment processed");

    } catch (TerminalException e) {
      // Cancellation detected - run compensation
      ctx.run(() -> refundPayment(order));
      throw e; // Re-throw to propagate cancellation
    }
  }
}

File: docs/code/go/cancellation-example.go
package example

import (
	restate "github.com/restatedev/sdk-go"
)

func processOrder(ctx restate.Context, order Order) error {
	// This executes - not an await point
	restate.Set(ctx, "status", "processing")

	// If cancelled before this await, Run won't execute
	// If cancelled during execution, then gets thrown after the execution
	payment, err := restate.Run(ctx, func(ctx restate.RunContext) (Payment, error) {
		return processPayment(order)
	})
	if err != nil {
		if restate.IsTerminalError(err) {
			// Cancellation detected - run compensation
			_, _ = restate.Run(ctx, func(ctx restate.RunContext) (any, error) {
				return nil, refundPayment(order)
			})
		}
		return err
	}

	// These one-way calls execute even if cancellation happened right before
	restate.Send(ctx, notificationService.Notify, order.UserID, "Payment processed")

	return nil
}
